### PR TITLE
[NVPTX] Don't use stack memory when bitcasting to/from v2i8

### PR DESCRIPTION
**Pull request by peterbell10**
`v2i8` is an unsupported type, so we hit the default legalization rules which perform the bitcast in stack memory and is very inefficient on GPU.

This adds a custom lowering where we pack `v2i8` into `i16` and from there use another bitcast node to reach the final desired type. And also the inverse unpacking `i16` into `v2i8`.

### CODE REVIEW

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 1]
```diff
@@ -551,6 +551,10 @@ NVPTXTargetLowering::NVPTXTargetLowering(const NVPTXTargetMachine &TM,
   setOperationAction(ISD::EXTRACT_VECTOR_ELT, MVT::v4i8, Custom);
   setOperationAction(ISD::INSERT_VECTOR_ELT, MVT::v4i8, Custom);
   setOperationAction(ISD::VECTOR_SHUFFLE, MVT::v4i8, Custom);
+
+  // Custom conversions to/from v2i8.
+  setOperationAction(ISD::BITCAST, MVT::v2i8, Custom);
```
Comment (justinfargnoli) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:556]: `Expand` doesn't work for this problem, right?

File: llvm/test/CodeGen/NVPTX/i8x2-instructions.ll [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+; CHECK-LABEL: test_trunc_2xi8(
+; CHECK:      ld.param.u32 [[R1:%r[0-9]+]], [test_trunc_2xi8_param_0];
+; CHECK:      mov.b32 {[[RS1:%rs[0-9]+]], [[RS2:%rs[0-9]+]]}, [[R1]];
+; CHECK:      shl.b16 	[[RS3:%rs[0-9]+]], [[RS2]], 8;
+; CHECK:      and.b16  [[RS4:%rs[0-9]+]], [[RS1]], 255;
+; CHECK:      or.b16   [[RS5:%rs[0-9]+]], [[RS4]], [[RS3]]
+; CHECK:      cvt.u32.u16  [[R2:%r[0-9]]], [[RS5]]
+; CHECK:      st.param.b32  [func_retval0], [[R2]];
+define i16 @test_trunc_2xi8(<2 x i16> %a) #0 {
+  %trunc = trunc <2 x i16> %a to <2 x i8>
```
Comment (justinfargnoli) [inline llvm/test/CodeGen/NVPTX/i8x2-instructions.ll:21]: Why include the `trunc` in this test and the `zext` in the test below?

**Review by justinfargnoli — COMMENTED** [review]
Overall, thanks for this PR. The current lowering is definitely not what we want! 

---

> we hit the default legalization rules which perform the bitcast in stack memory

Do you know where this default legalization rule is implemented? 

---

Note: [`mov` vector-to-scalar (pack) or scalar-to-vector (unpack)](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-mov-2) doesn't support `.b8`.

File: llvm/test/CodeGen/NVPTX/i8x2-instructions.ll [hunk 3]
```diff
@@ -0,0 +1,36 @@
+; RUN: llc < %s -mtriple=nvptx64-nvidia-cuda -mcpu=sm_90 -mattr=+ptx80 -asm-verbose=false \
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply llvm/test/CodeGen/NVPTX/i8x2-instructions.ll:1]: I haven't used that before but I gave it a shot and it didn't generate any checks at all for some reason, perhaps I was doing something wrong. Not sure.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 1 above)
Comment (peterbell10) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:556]: No it just results in the same stack usage.

File: llvm/test/CodeGen/NVPTX/i8x2-instructions.ll (hunk 3 above)
[the remark this answers is not in the document]
Comment (justinfargnoli) [reply llvm/test/CodeGen/NVPTX/i8x2-instructions.ll:1]: I think you need to delete all the existing `CHECK` statements first. 

If that doesn't work (and the rest of the MR looks good), I'll just submit a follow-up patch that auto-generates the test.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 4]
```diff
@@ -2311,6 +2315,47 @@ NVPTXTargetLowering::LowerCONCAT_VECTORS(SDValue Op, SelectionDAG &DAG) const {
...
+
+  if (ToVT == MVT::v2i8) {
+    // Bitcast to i16 and unpack elements into a vector
+    SDValue AsInt = maybeBitcast(MVT::i16, Op->getOperand(0));
+    SDValue Vec0 = DAG.getNode(ISD::TRUNCATE, DL, MVT::i8, AsInt);
+    SDValue Const8 = DAG.getConstant(8, DL, MVT::i16);
+    SDValue Vec1 =
+        DAG.getNode(ISD::TRUNCATE, DL, MVT::i8,
+                    DAG.getNode(ISD::SRL, DL, MVT::i16, {AsInt, Const8}));
+    return DAG.getNode(ISD::BUILD_VECTOR, DL, MVT::v2i8, {Vec0, Vec1});
```
Comment (justinfargnoli) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:2341]: This snippet is only used by the call in `ReplaceNodeResults()`. The `else if` snippet below is only used by the call in `LowerOperation()`. 

Let's split these cases into separate methods: 
- `ReplaceBitcast()` for the `if (ToVT == MVT::v2i8)` snippet.
- `LowerBITCAST()` for the `else if (FromVT == MVT::v2i8)` snippet.

Comment (peterbell10) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: Done

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 5]
```diff
@@ -2311,6 +2322,31 @@ NVPTXTargetLowering::LowerCONCAT_VECTORS(SDValue Op, SelectionDAG &DAG) const {
...
 
+SDValue NVPTXTargetLowering::LowerBITCAST(SDValue Op, SelectionDAG &DAG) const {
+  // Handle bitcasting from v2i8 without hitting the default promotion
+  // strategy which goes through stack memory.
+  SDLoc DL(Op);
+
+  EVT ToVT = Op->getValueType(0);
+  EVT FromVT = Op->getOperand(0)->getValueType(0);
+
+  if (FromVT == MVT::v2i8) {
```
Comment (justinfargnoli) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:2333]: Prefer early returns: https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 6]
```diff
@@ -6128,6 +6166,26 @@ SDValue NVPTXTargetLowering::PerformDAGCombine(SDNode *N,
...
 
+static void ReplaceBITCAST(SDNode *Node, SelectionDAG &DAG,
+                           SmallVectorImpl<SDValue> &Results) {
+  // Handle bitcasting to v2i8 without hitting the default promotion
+  // strategy which goes through stack memory.
+  SDValue Op(Node, 0);
+  SDLoc DL(Node);
+
+  EVT ToVT = Op->getValueType(0);
+  if (ToVT == MVT::v2i8) {
```
Comment (justinfargnoli) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6177]: Prefer early returns: https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

**Review by justinfargnoli — APPROVED** [review]
LGTM, pending the resolution of all conversations!

**Pull request merged**